Implements volumetric (5d) affine grid generation.

elistevens opened the pull request:
I've implemented affine grid generation for volumetric (5d) inputs. The implementation is based off of the spatial implementation, extended by one dimension. I have a few questions about my implementation vs. the existing one that I will add inline.

I have some extensive test cases for the forward pass here: https://gist.github.com/elistevens/6e3bfb20d8d0652b83bd16b3e911285b However, they use `pytest.fixture` extensively, so I'm not sure the best way to incorporate them into the pytorch test suite. Suggestions? I have not tested backwards at all.

Diff probably best viewed with whitespace changes ignored.

Thanks for considering!

elistevens committed:
Implements volumetric (5d) affine grid generation.

elistevens commented on torch/nn/_functions/vision.py:
```diff
@@ -93,35 +93,74 @@ def _enforce_cudnn(input):
+
+            base_grid[:, :, :, :, 0] = w_points
+            base_grid[:, :, :, :, 1] = h_points
+            base_grid[:, :, :, :, 2] = d_points
```
I was surprised that the 2d equivalent to `w_points` was assigned to `base_grid[:, :, :, :, 0]` instead of `base_grid[...,1]`.

I think that this is explained by the first Note at https://pytorch.org/docs/master/nn.html#torch.nn.functional.grid_sample :

> To avoid confusion in notation, let’s note that x corresponds to the width dimension IW, y corresponds to the height dimension IH and z corresponds to the depth dimension ID.

But I'd like someone else to confirm that this is correct.

elistevens commented on torch/nn/_functions/vision.py:
```diff
@@ -93,35 +93,74 @@ def _enforce_cudnn(input):
-        base_grid[:, :, :, 1] = torch.ger(linear_points, torch.ones(W)).expand_as(base_grid[:, :, :, 1])
-        base_grid[:, :, :, 2] = 1
-        ctx.base_grid = base_grid
-        grid = torch.bmm(base_grid.view(N, H * W, 3), theta.transpose(1, 2))
```
This implementation seems cleaner to me than the original, but I have not done performance testing. Should I backport this style to the 2d case?

elistevens commented on torch/nn/_functions/vision.py:
```diff
@@ -93,35 +93,74 @@ def _enforce_cudnn(input):
+            if theta.is_cuda:
+                AffineGridGenerator._enforce_cudnn(theta)
+                assert False
+            ctx.is_cuda = False
```
This `if theta.is_cuda:` check seems needlessly restrictive. The `def affine_grid_generator(theta, size):` function handles the dispatch between CPU and CUDA; why repeat the check here as well?

SsnL commented on torch/nn/_functions/vision.py:
```diff
@@ -93,35 +93,74 @@ def _enforce_cudnn(input):
-        if grad_grid.is_cuda:
-            AffineGridGenerator._enforce_cudnn(grad_grid)
+        if len(ctx.size) == 5:
+            N, C, D, H, W = ctx.size
```
There seem to be a lot of redundant code between these two `if` cases. Can you make them share part of it?

elistevens replied:
I'm happy to consolidate. I originally didn't to make it clear that the 2D case hadn't meaningfully changed.

elistevens replied:
I see three lines in common:

        assert ctx.is_cuda == grad_grid.is_cuda
        base_grid = ctx.base_grid
        # ...
        grad_theta = grad_theta.transpose(1, 2)

I'll pull those out. I don't think that the other similarities are worth trying to de-duplicate, since the code would get significantly less readable.

ssnl reviewed (changes requested):
Can you add tests for this too? And FYI this function should be eventually move to ATen c++ side one day, and the cudnn check logic will be a lot simpler when that happens.

yf225 commented:
@elistevens Let us know if you have time to address @SsnL 's comments

elistevens commented:
Sorry for the slow response. I've been traveling and will be for another
week still, but I intend to work on this when I get back.

On Tue, Jul 10, 2018, 09:50 Will Feng <notifications@github.com> wrote:

> @elistevens <https://github.com/elistevens> Let us know if you have time
> to address @SsnL <https://github.com/SsnL> 's comments
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/8322#issuecomment-403891417>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAIbIM1f9phgRQPh8qwHNDt2wMWsgDi6ks5uFNs4gaJpZM4Uh4b5>
> .
>

DTechSolutions commented on torch/nn/_functions/vision.py:
```diff
@@ -93,35 +93,74 @@ def _enforce_cudnn(input):
-        grid = grid.view(N, H, W, 2)
+
+        if len(size) == 5:
+            N, C, D, H, W = size
```
shouldn't this be, D at the end?
`N, C, H, W, D = size`

DTechSolutions replied:
shouldn't this be, D at the end?
`N, C, H, W, D = size`

elistevens committed:
Added affine tests. Test style is adapted from pytest fixtures.

elistevens committed:
Clean up AffineGridGenerator implementation. Removes overly-restrictive checks on CuDNN etc.

elistevens commented:
@SsnL @yf225 I've added tests and cleaned up the implementation.

Note that I've removed all of the "if `theta.is_cuda` then we must use CuDNN" checks. The `affine_grid_generator` does CuDNN dispatch, and if the control flow gets to `AffineGridGenerator` at all, it will happily work on CPU or CUDA tensors without complaint (while not implemented ATM, this would make it easier to verify that the python version outputs match the CuDNN outputs).

Are there any other changes desired?

elistevens commented:
The jenkins builds are failing, though the tests work locally.  There are going to be some extra pushes while I try and figure out what the difference is between my local setup and the formal testing. Sorry for the noise.

elistevens committed:
Fix lint errors, prints scipy version.

elistevens committed:
Update scipy version, more flake8 fixes.

elistevens commented:
@soumith Yes, absolutely. This was intended to be a temporary measure to get around the older version issue. I'll revert it once the docker images are updated.  :)

yf225 commented:
@elistevens I just updated the scipy version in our docker images to 1.1.0. Do you mind reverting the change to requirements.txt and see if CI passes?

elistevens commented:
Will do tonight. Thanks!

On Mon, Jul 23, 2018, 12:16 Will Feng <notifications@github.com> wrote:

> @elistevens <https://github.com/elistevens> I just updated the scipy
> version in our docker images to 1.1.0. Do you mind reverting the change to
> requirements.txt and see if CI passes?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/8322#issuecomment-407170066>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAIbIMTkQVNulr40F5hSHuhxYXRYCk9Cks5uJiEKgaJpZM4Uh4b5>
> .
>

elistevens committed:
Remove scipy requirement, replace @ with np.dot(...).

elistevens committed:
Adds py2 from future import division

elistevens commented:
In case anyone is wondering about the `RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88` warnings, they're not actually a problem according to: https://stackoverflow.com/questions/40845304/runtimewarning-numpy-dtype-size-changed-may-indicate-binary-incompatibility

https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-trusty-pynightly-test/11964/console doesn't have scipy installed.

https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-build/13558/console
```
23:14:25    Creating library build\temp.win-amd64-3.6\Release\torch/csrc\_C.cp36-win_amd64.lib and object build\temp.win-amd64-3.6\Release\torch/csrc\_C.cp36-win_amd64.exp
23:14:25 Size.obj : error LNK2001: unresolved external symbol "struct torch::jit::Value * __cdecl torch::jit::insertConstant(struct torch::jit::Graph &,struct torch::jit::IValue,class at::optional<struct torch::jit::script::SourceRange>)" (?insertConstant@jit@torch@@YAPEAUValue@12@AEAUGraph@12@UIValue@12@V?$optional@USourceRange@script@jit@torch@@@at@@@Z)
23:14:25 build\lib.win-amd64-3.6\torch\_C.cp36-win_amd64.pyd : fatal error LNK1120: 1 unresolved externals
23:14:25 error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Community\\VC\\Tools\\MSVC\\14.11.25503\\bin\\HostX86\\x64\\link.exe' failed with exit status 1120
```
This doesn't seem to be related to my branch. Will this test failing be a merge blocker?

I think the rest of the issues are due to integer division under python 2. Testing now.

elistevens committed:
Merge branch 'master' into affine_grid_volumetric_support

elistevens commented:
The last build had a handful of errors that all seemed unrelated to my changes (missing scipy on some of the testing images, windows build error, etc.).

@SsnL Are there any other desired changes?

elistevens commented:
@yf225 the remaining three failures are all due to scipy being missing entirely from the relevant images. Is that something you care to fix?

yf225 commented:
@elistevens We should probably use the `TEST_SCIPY` flag (https://github.com/pytorch/pytorch/blob/master/test/common.py#L84) to gate the tests that require scipy in test_affine.py.

elistevens committed:
Added TEST_SCIPY etc. guards.

elistevens commented:
@SsnL @yf225 any further changes desired?

SsnL commented on test/test_affine.py:
```diff
@@ -0,0 +1,602 @@
+from __future__ import print_function, division
```
Can you fold this in test_nn and remove the comments/logging please?

soumith commented:
@elistevens after Simon's comment is addressed, this PR is good to go, i.e.:

> Can you fold this in test_nn and remove the comments/logging please?

elistevens committed:
Move TestAffine to test_nn.py

soumith commented on test/run_test.py:
```diff
@@ -1,345 +1,345 @@
-#!/usr/bin/env python
```
it looks like you changed something in this file that you shouldn't have.
I looked, and it's whitespace changes.

Can you revert changes to this file entirely.

elistevens replied:
Whoops, yeah. My editor is usually good about preserving line endings and such, but I upgraded versions yesterday, so perhaps some settings got changed.

I'll revert in a moment.

elistevens committed:
Correct unintended whitespace change to run_test.py

Keeps the 'affine' line removal, since that's still needed.

elistevens committed:
Merge branch 'master' into affine_grid_volumetric_support

elistevens committed:
Fixing line-too-long error.

SsnL commented on test/test_nn.py:
```diff
@@ -8347,6 +8351,392 @@ def __call__(self, input):
+    return transform_tensor, transform_ary, grid_ary
+
+
+class TestAffine(TestCase):
```
Can this just be a method of `TestNN` instead of a new method?

elistevens replied:
If that's preferred, there's no technical reason preventing it.

I personally find the "one giant file, one giant class" approach to test cases hard to use, since it makes it more difficult to run the subset of tests that I'm interested in, but if that's how you all prefer to structure things, I don't need to bikeshed about it.

SsnL replied:
Thanks!

elistevens committed:
Move affine tests into TestNN
